### PR TITLE
Feat/search box

### PR DESCRIPTION
**Pull request by tmaciulis22**
Created reusable SearchBox component, which overlays its search results on other ui elements.
Also added an OutsideClickListener wrapper component, which listens for clicks outside of wrapped element, and invokes a callback when clicked outside of it.
The search box appers when clicked on **Add new member** icon.
image
image
image

### CODE REVIEW

**Comment by tmaciulis22** [comment]
[Task 4: Create search box and "Add" button for searching and adding people to group](https://trello.com/c/Wn5H4Ra2/53-task-4-create-search-box-and-add-button-for-searching-and-adding-people-to-group)

File: billchop-fe/src/backend/server-config.json [hunk 1]
```diff
@@ -1,4 +1,3 @@
 {
-  "host": "https://localhost:5001",
-  "createTestUser": true
```
Comment (AinoraZ) [inline billchop-fe/src/backend/server-config.json:3]: Remove this from ServerConfig.ts as well

File: billchop-fe/src/components/GroupPageHeader.tsx [hunk 2]
```diff
@@ -1,58 +1,119 @@
...
   IGroupPageHeaderProps,
   IGroupPageHeaderState
-> {
+  > {
   constructor(props: IGroupPageHeaderProps) {
     super(props);
     this.state = {
       showAddBillModal: false,
+      showSearchBox: false,
+      foundUsers: [],
```
Comment (AinoraZ) [inline billchop-fe/src/components/GroupPageHeader.tsx:32]: [No Fix]
This and all search UI logic should probably be placed in its own sub-component at some point. Its fine for now, could be a refactoring task for next iteration.

File: billchop-fe/src/components/SearchBox.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+    const { onChange } = this.props;
+    onChange(inputValue);
+  };
+
+  delayedHandleInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
+    const inputValue = event.target.value;
+    this.setState({ inputValue });
+    const delayed = debounce(() => this.handleInputChange(inputValue), 300);
+    delayed();
```
Comment (AinoraZ) [inline billchop-fe/src/components/SearchBox.tsx]: Does this work? I'm not 100% familiar with how debounced works, perhaps it has some global state, but to me this looks like you are creating a new debounced version of the function every time, which means it's probably not debounced. Try increasing the time to something bigger (10s?) and see if it works.

To make debounced work, I think you will need to do something like this:
```typescript
  handleInputChange = (inputValue: string): void => {
    if (!inputValue) return;

    const { onChange } = this.props;
    onChange(inputValue);
  };

  handleDebouncedInputChange = debounce(this.handleInputChange, 300);

  delayedHandleInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const inputValue = event.target.value;
    this.setState({ inputValue });

    this.handleDebouncedInputChange(inputValue);
  };
```

**Review by AinoraZ — CHANGES_REQUESTED** [review]

**Review by AinoraZ — APPROVED** [review]

**Pull request merged**